### PR TITLE
Make always on top window resizable

### PR DESCRIPTION
**Pull request by horymury**
- minimum size is the starting size of the window
- aspect ratio of video gets preserved on resize, and empty spaces either on top-bottom/left-right are gray background. this is discussed and agreed with PM

### CODE REVIEW

**Comment by hristoterezov** [comment]
@horymury I noticed the following behaviour:

1. Every time when the always on top window is reopened (even during the same call) the size is reset to the default one.
2. The window can be resized to a different aspect ratio. We don't have any limitations for the aspect ratio.

I just want to make sure that this is not a problem.

Otherwise looks good to me and if you confirm that the above is the desired behaviour I'll merge the PR.

**Comment by horymury** [comment]
@hristoterezov Emil had the following requests:

1. see if we can find out why AOT window keeps growing on moving it - Windows only
2. keep the last set size on the same meeting session or even the same VOD session
3. keep the aspect ratio on resize

File: alwaysontop/render.js [hunk 1]
```diff
@@ -234,19 +263,38 @@ class AlwaysOnTop extends EventEmitter {
...
+                previousHeight = size[1];
+            },
             onbeforeunload: () => {
                 this.emit(ALWAYSONTOP_WILL_CLOSE);
                 this._api.removeListener(
                     'largeVideoChanged',
                     this._updateLargeVideoSrc
                 );
             },
+            onresize: debounce(() => {
```
Comment (hristoterezov) [inline alwaysontop/render.js]: Maybe for Mac you can use https://electronjs.org/docs/api/browser-window#winsetaspectratioaspectratio-extrasize-macos

Comment (hristoterezov) [inline alwaysontop/render.js]: Why do we need the debounce?

File: alwaysontop/render.js [hunk 2]
```diff
@@ -304,12 +352,17 @@ class AlwaysOnTop extends EventEmitter {
                 x: position[0],
                 y: position[1]
             };
+
+            const size = this._alwaysOnTopBrowserWindow.getSize();
+            const width = size[0];
+            const height = size[1];
+            this._sendSize({ width, height });
```
Comment (hristoterezov) [inline alwaysontop/render.js]:
```
this._sendSize({
    width: size[0],
    height: size[1]
})
```

Isn't this looking better?

File: alwaysontop/render.js [hunk 3]
```diff
@@ -146,6 +150,30 @@ class AlwaysOnTop extends EventEmitter {
         }
     }
 
+    _sendResetSize() {
```
Comment (hristoterezov) [inline alwaysontop/render.js]: Missing JSDoc.

File: alwaysontop/render.js [hunk 4]
```diff
@@ -146,6 +150,30 @@ class AlwaysOnTop extends EventEmitter {
...
+        ipcRenderer.send('jitsi-always-on-top', {
+            type: 'event',
+            data: {
+                name: 'size',
+                type: 'reset',
+            }
+        });
+    }
+
+    _sendSize({ width, height }) {
```
Comment (hristoterezov) [inline alwaysontop/render.js]: Missing JSDoc.

File: alwaysontop/render.js [hunk 5]
```diff
@@ -1,12 +1,16 @@
...
 const { ipcRenderer, remote } = require('electron');
+const debounce = require('debounce');
 
 const { EventEmitter } = require('events');
 const os = require('os');
 const path = require('path');
 
 const { ALWAYSONTOP_WILL_CLOSE } = require('./constants');
 
+const aspectRatio = 1.7777777777777777;
```
[the remark this answers is not in the document]
Comment (hristoterezov) [reply alwaysontop/render.js]: Also please add JSDoc.

File: alwaysontop/render.js [hunk 6]
```diff
@@ -1,12 +1,16 @@
...
+let previousWidth;
```
Comment (hristoterezov) [inline alwaysontop/render.js]: Missing JSDoc.

File: alwaysontop/render.js [hunk 7]
```diff
@@ -1,12 +1,16 @@
...
+let previousHeight;
```
Comment (hristoterezov) [inline alwaysontop/render.js]: Missing JSDoc.

File: alwaysontop/render.js [hunk 8]
```diff
@@ -1,12 +1,16 @@
...
 /**
```
Comment (hristoterezov) [inline alwaysontop/render.js]: Add new line

**Comment by horymury** [comment]
> Left some comments and questions!
> 
> Also when I tested it the resize feels a little jumpy and choppy. I think there is a problem with the calculation logic. Could you please take a look?

The choppines and jumpiness where resolved by moving the logic to main process.

File: alwaysontop/render.js (hunk 1 above)
Comment (horymury) [reply alwaysontop/render.js]: removed debounce

Comment (horymury) [reply alwaysontop/render.js]: I'm working on windows so it's a bit tricky to test it out. But with the changes I made it should work just fine with the current logic on MAC as well

File: alwaysontop/main.js [hunk 9]
```diff
@@ -38,27 +56,55 @@ function onAlwaysOnTopWindow(
...
                 show: false
-            }, getPosition())
+            }, getPosition(), getSize())
         );
         win.once('ready-to-show', () => {
             win.showInactive();
         });
+
+        //for macOS we use the built-in setAspectRatio on resize, for other we use custom implementation
+        if (os.type() === 'Darwin') {
```
Comment (hristoterezov) [inline alwaysontop/main.js]: Why did we move the resize code from the render to the main process?

File: alwaysontop/main.js [hunk 10]
```diff
@@ -38,27 +56,55 @@ function onAlwaysOnTopWindow(
...
+            win.setAspectRatio(16/9);
+        } else {
+            win.on('will-resize', (e, newBounds) => {
+                oldSize = win.getSize();
+                const mousePos = robot.getMousePos();
```
Comment (hristoterezov) [inline alwaysontop/main.js]: Do we really need to use robotjs for this? Isn't it https://github.com/electron/electron/blob/master/docs/api/screen.md#screengetcursorscreenpoint do the job?

File: alwaysontop/main.js [hunk 11]
```diff
@@ -38,27 +56,55 @@ function onAlwaysOnTopWindow(
...
+                const windowBottomRightPos = {
+                    x: newBounds.x + newBounds.width - 16,
```
Comment (hristoterezov) [inline alwaysontop/main.js]: Why `-16`?

File: alwaysontop/main.js [hunk 12]
```diff
@@ -38,27 +56,55 @@ function onAlwaysOnTopWindow(
...
+                    y: newBounds.y + newBounds.height - 16,
+                };
+                //prevent resize from bottom right corner as it is buggy.
+                if (mousePos.x >= windowBottomRightPos.x && mousePos.y >= windowBottomRightPos.y) {
```
Comment (hristoterezov) [inline alwaysontop/main.js]: I'm wondering isn't it there a better way to detect this? 

I'm worried that you can end up in the same situation even if you started the resize from another corner. Although probably we shouldn't care about this.

File: alwaysontop/main.js [hunk 13]
```diff
@@ -1,19 +1,37 @@
...
 const SIZE = {
     width: 320,
     height: 180
 };
 
+/**
+ * The aspect ratio to preserve during AOT window resize
+ * @type {number}
+ */
+const aspectRatio = 16 / 9;
```
[the remark this answers is not in the document]
Comment (horymury) [reply alwaysontop/main.js]: done

File: alwaysontop/main.js (hunk 9 above)
Comment (horymury) [reply alwaysontop/main.js]: as it seemed to be more fluent on main process

File: alwaysontop/main.js (hunk 11 above)
Comment (horymury) [reply alwaysontop/main.js]: I've set this as the diagonal resize does not necessarily start precisely on the window's right bottom corner, it has some pixels threshold on right and bottom window edges in which is still a diagonal resize

File: alwaysontop/main.js (hunk 10 above)
Comment (horymury) [reply alwaysontop/main.js]: Please see following image
First is robotjs detection, second is electron screen detection on a monitor having scale 125%. 
Please note that robotjs correctly shows the y axe as 0 in this case, and electron screen detects it wrongly as 200 something. I think this wrong detection electron-side causes all the bugs reported on Electron related to scaled monitors on Windows.

image

File: alwaysontop/main.js (hunk 12 above)
Comment (horymury) [reply alwaysontop/main.js]: Not sure if it s possible to end up in this situation from another corner, as the check is done on future position of the window,  and the future position doesn't seem possible to have the bottom right corner 'upper' and 'lefter' from current mouse position :)

**Pull request merged**